# HG changeset patch
# User Olaf Till <olaf.till@uni-jena.de>
# Date 1287622293 14400
#      Wed Oct 20 20:51:33 2010 -0400
# Node ID 134d357e1c035edf8cd3ef34aa677a6229c5f9dc
# Parent  a44f979a35ce4a5bf6c8eee4f19f744a398455e3
optimization/optimset.m: Use cell2struct instead of struct to handle cell arrays as options

diff --git a/scripts/ChangeLog b/scripts/ChangeLog
--- a/scripts/ChangeLog
+++ b/scripts/ChangeLog
@@ -1,8 +1,13 @@
+2010-10-20  Olaf Till  <olaf.till@uni-jena.de>
+
+	* optimization/optimset.m: Use cell2struct instead of struct to
+	handle cell arrays as options.
+
 2010-10-20  John W. Eaton  <jwe@octave.org>
 
 	* io/strread.m, optimization/sqp.m, plot/daspect.m,
 	plot/pbaspect.m, plot/gnuplot_drawnow.m, plot/isocolors.m,
 	plot/isonormals.m, plot/legend.m, plot/print.m,
 	plot/private/__interp_cube__.m, plot/__fltk_print__.m,
 	plot/__gnuplot_print__.m, plot/__print_parse_opts__.m,
 	statistics/base/quantile.m: Style fixes.
diff --git a/scripts/optimization/optimset.m b/scripts/optimization/optimset.m
--- a/scripts/optimization/optimset.m
+++ b/scripts/optimization/optimset.m
@@ -71,21 +71,23 @@ function retval = optimset (varargin)
           warning ("unrecognized option: %s", key);
         endif
       endif
       old.(key) = val;
     endfor
     retval = old;
   elseif (rem (nargs, 2) && isstruct (varargin{1}))
     ## Set values in old from name/value pairs.
-    retval = optimset (varargin{1}, struct (varargin{2:end}));
+    pairs = reshape (varargin(2:end), 2, []);
+    retval = optimset (varargin{1}, cell2struct (pairs(2, :), pairs(1, :), 2));
   elseif (rem (nargs, 2) == 0)
     ## Create struct.  Default values are replaced by those specified by
     ## name/value pairs.
-    retval = optimset (struct (), struct (varargin{:}));
+    pairs = reshape (varargin, 2, []);
+    retval = optimset (struct (), cell2struct (pairs(2, :), pairs(1, :), 2));
   else
     print_usage ();
   endif
 
 endfunction
 
 %!assert (optimget (optimset ('tolx', 1e-2), 'tOLx'), 1e-2)
 %!assert (isfield (optimset ('tolFun', 1e-3), 'TolFun'))
